src/memtest.rs

```diff
@@ -0,0 +1,456 @@
+        fmt,
+        ptr::{read_volatile, write_bytes, write_volatile},
+    },
+};
+
+// TODO: Intend to convert this module to a standalone `no_std` crate
+// TODO: TimeoutChecker will be a trait instead
+
+#[derive(Debug)]
+pub enum MemtestOutcome {
+    Pass,
+    Fail(MemtestFailure),
+}
+
+#[derive(Debug)]
+pub enum MemtestFailure {
+    /// Failure due to the actual value read being different from the expected value
+    UnexpectedValue {
+        address: usize,
+        expected: usize,
+        actual: usize,
+    },
+    /// Failure due to the two memory locations being compared returning two different values
+    /// This is used by tests where memory is splitted in two and random values are written in pairs
+    MismatchedValues {
+        address1: usize,
+        value1: usize,
+        address2: usize,
+        value2: usize,
+    },
+}
+
+#[derive(Debug)]
+pub enum MemtestError {
+    Timeout,
+    Other(anyhow::Error),
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum MemtestType {
+    TestOwnAddress,
+    TestRandomVal,
+    TestXor,
+    TestSub,
+    TestMul,
+    TestDiv,
+    TestOr,
+    TestAnd,
+    TestSeqInc,
+    TestSolidBits,
+    TestCheckerboard,
+    TestBlockSeq,
+}
+
+/// Write the address of each memory location to itself
+/// then read back the value and verify that it matches the expected address
+pub unsafe fn test_own_address(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestOwnAddress");
+    // TODO: According to the linux memtester, this needs to be run several times,
+    //       and with alternating complements of address
+    let expected_iter = u64::try_from(count.checked_mul(2).context("expected_iter overflowed")?)
+        .context("Failed to convert expected_iter to u64")?;
+    timeout_checker.init(expected_iter);
+
+    for i in 0..count {
+        timeout_checker.check()?;
+        let ptr = base_ptr.add(i);
+        write_volatile(ptr, ptr as usize);
+    }
+
+    for i in 0..count {
+        timeout_checker.check()?;
+        let ptr = base_ptr.add(i);
+        let val = read_volatile(ptr);
+        if val != ptr as usize {
+            info!("Test failed at {ptr:?}");
+            return Ok(MemtestOutcome::Fail(MemtestFailure::UnexpectedValue {
+                address: ptr as usize,
+                expected: ptr as usize,
+                actual: val,
+            }));
+        }
+    }
+    Ok(MemtestOutcome::Pass)
+}
+
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// write a random value for each pair of memory locations
+/// read and compare the two halves of the memory region
+pub unsafe fn test_random_val(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestRandomVal");
+    let half_count = count / 2;
```

Comment:
In all of these two-region tests, you should probably return an error if `count < 2`. Otherwise, the test will always pass because it never actually tested anything...